File: cpp/src/ch04/random_action.cc
#include <random>
#include "random_action.h"

auto mt_for_action_2 = std::mt19937(0);

State random_action(const State &state)
{
    State now_state = state;
    for (int character_id = 0; character_id < CHARACTER_N; character_id++)
    {
        int y = mt_for_action_2() % H;
        int x = mt_for_action_2() % W;

        now_state.set_character(character_id, y, x);
    }
    return now_state;
}

File: rust/src/ch04/random_action.rs
use crate::ch04::maze_state;

use rand::{thread_rng, Rng};

// seems unnecessary but keep the function signatures the same
pub fn random_action_factory() -> Box<maze_state::ActionFunc> {
    Box::new(move |state| random_action(state))
}

fn random_action(
    initial_state: &maze_state::AutoMoveMazeState,
) -> maze_state::AutoMoveMazeState {
    let mut rng = thread_rng();
    let mut state = initial_state.clone();
    for id in 0..state.params.num_characters {
        let y = rng.gen_range(0..state.params.height);
        let x = rng.gen_range(0..state.params.width);

        state.set_character(id, y, x);
    }

    state
}
